### ISSUE TITLE
Maximum recursion depth exceed for self referential collections

### ISSUE THREAD

**Issue by lucashowell** [thomwiggers/django-mongoengine-forms#10]
I'm trying to upgrade an old project from django-mongodbforms to use django-mongoengine-forms, however am encountering an error that I previously had with mongodbforms (that they fixed).  The previous error was this:

https://github.com/jschrewe/django-mongodbforms/issues/60

It has to do with collections that have self-referential foreign keys, I believe (when I comment out this foreign key, the problem goes away).  Example collection:

```
class Group(Document):
    parent_group = ReferenceField("Group")
    name = StringField(required=True, unique=True)
```

```
Unhandled exception in thread started by <function wrapper at 0x10eecf488>
Traceback (most recent call last):
  File "/Users/lucas/.virtualenvs/test_dpl/lib/python2.7/site-packages/django/utils/autoreload.py", line 226, in wrapper
    fn(*args, **kwargs)
  File "/Users/lucas/.virtualenvs/test_dpl/lib/python2.7/site-packages/django/core/management/commands/runserver.py", line 121, in inner_run
    self.check(display_num_errors=True)
  File "/Users/lucas/.virtualenvs/test_dpl/lib/python2.7/site-packages/django/core/management/base.py", line 374, in check
    include_deployment_checks=include_deployment_checks,
  File "/Users/lucas/.virtualenvs/test_dpl/lib/python2.7/site-packages/django/core/management/base.py", line 361, in _run_checks
    return checks.run_checks(**kwargs)
  File "/Users/lucas/.virtualenvs/test_dpl/lib/python2.7/site-packages/django/core/checks/registry.py", line 81, in run_checks
    new_errors = check(app_configs=app_configs)
  File "/Users/lucas/.virtualenvs/test_dpl/lib/python2.7/site-packages/django/core/checks/urls.py", line 14, in check_url_config
    return check_resolver(resolver)
  File "/Users/lucas/.virtualenvs/test_dpl/lib/python2.7/site-packages/django/core/checks/urls.py", line 24, in check_resolver
    for pattern in resolver.url_patterns:
  File "/Users/lucas/.virtualenvs/test_dpl/lib/python2.7/site-packages/django/utils/functional.py", line 35, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/Users/lucas/.virtualenvs/test_dpl/lib/python2.7/site-packages/django/urls/resolvers.py", line 313, in url_patterns
    patterns = getattr(self.urlconf_module, "urlpatterns", self.urlconf_module)
  File "/Users/lucas/.virtualenvs/test_dpl/lib/python2.7/site-packages/django/utils/functional.py", line 35, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/Users/lucas/.virtualenvs/test_dpl/lib/python2.7/site-packages/django/urls/resolvers.py", line 306, in urlconf_module
    return import_module(self.urlconf_name)
  File "/usr/local/Cellar/python/2.7.11/Frameworks/Python.framework/Versions/2.7/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "/Users/lucas/timberline/nexleaf/dataportal/NexleafDP/NexleafDP/dataportal/dataportal/urls.py", line 30, in <module>
    url(r'^accounts/', include('dataportal.accounts.urls')),
  File "/Users/lucas/.virtualenvs/test_dpl/lib/python2.7/site-packages/django/conf/urls/__init__.py", line 50, in include
    urlconf_module = import_module(urlconf_module)
  File "/usr/local/Cellar/python/2.7.11/Frameworks/Python.framework/Versions/2.7/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "/Users/lucas/timberline/nexleaf/dataportal/NexleafDP/NexleafDP/dataportal/dataportal/accounts/urls.py", line 10, in <module>
    from dataportal.accounts import views as account_views
  File "/Users/lucas/timberline/nexleaf/dataportal/NexleafDP/NexleafDP/dataportal/dataportal/accounts/views.py", line 22, in <module>
    from dataportal.accounts.forms import UpdateProfileForm, PasswordChangeForm, \
  File "/Users/lucas/timberline/nexleaf/dataportal/NexleafDP/NexleafDP/dataportal/dataportal/accounts/forms.py", line 104, in <module>
    class UpdateProfileForm(DocumentForm):
  File "/Users/lucas/.virtualenvs/test_dpl/lib/python2.7/site-packages/mongodbforms/documents.py", line 322, in __new__
    formfield_generator)
  File "/Users/lucas/.virtualenvs/test_dpl/lib/python2.7/site-packages/mongodbforms/documents.py", line 248, in fields_for_document
    formfield = field_generator.generate(f, **kwargs)
  File "/Users/lucas/.virtualenvs/test_dpl/lib/python2.7/site-packages/mongodbforms/fieldgenerator.py", line 403, in generate
    return sup.generate(field, **kwargs)
  File "/Users/lucas/.virtualenvs/test_dpl/lib/python2.7/site-packages/mongodbforms/fieldgenerator.py", line 81, in generate
    return getattr(self, attr_name)(field, **kwargs)
  File "/Users/lucas/.virtualenvs/test_dpl/lib/python2.7/site-packages/mongodbforms/fieldgenerator.py", line 307, in generate_referencefield
    'queryset': field.document_type.objects.clone(),
  File "/Users/lucas/.virtualenvs/test_dpl/lib/python2.7/site-packages/mongoengine/queryset/manager.py", line 36, in __get__
    queryset_class = owner._meta.get('queryset_class', self.default)
  File "/Users/lucas/.virtualenvs/test_dpl/lib/python2.7/site-packages/django/utils/functional.py", line 234, in inner
    self._setup()
  File "/Users/lucas/.virtualenvs/test_dpl/lib/python2.7/site-packages/mongodbforms/documentoptions.py", line 75, in _setup
    self._wrapped = DocumentMetaWrapper(self._document, self._meta)
  File "/Users/lucas/.virtualenvs/test_dpl/lib/python2.7/site-packages/mongodbforms/documentoptions.py", line 140, in __init__
    self._setup_document_fields()
  File "/Users/lucas/.virtualenvs/test_dpl/lib/python2.7/site-packages/mongodbforms/documentoptions.py", line 186, in _setup_document_fields
    LazyDocumentMetaWrapper)) and
  File "/Users/lucas/.virtualenvs/test_dpl/bin/../lib/python2.7/abc.py", line 131, in __instancecheck__
    subclass = getattr(instance, '__class__', None)
  File "/Users/lucas/.virtualenvs/test_dpl/lib/python2.7/site-packages/django/utils/functional.py", line 234, in inner
    self._setup()
  File "/Users/lucas/.virtualenvs/test_dpl/lib/python2.7/site-packages/mongodbforms/documentoptions.py", line 75, in _setup
    self._wrapped = DocumentMetaWrapper(self._document, self._meta)
  File "/Users/lucas/.virtualenvs/test_dpl/lib/python2.7/site-packages/mongodbforms/documentoptions.py", line 140, in __init__
    self._setup_document_fields()
  File "/Users/lucas/.virtualenvs/test_dpl/lib/python2.7/site-packages/mongodbforms/documentoptions.py", line 186, in _setup_document_fields
    LazyDocumentMetaWrapper)) and
  File "/Users/lucas/.virtualenvs/test_dpl/bin/../lib/python2.7/abc.py", line 131, in __instancecheck__
    subclass = getattr(instance, '__class__', None)
  ... etc...

RuntimeError: maximum recursion depth exceeded while calling a Python object
```

**Comment by thomwiggers** [comment]
I've added this test, but it doesn't reproduce your issue. (To be fair, I didn't really try very hard) Of course, it doesn't do anything exciting either. Could you elaborate, and perhaps come up with a test case?

I should also point out that I have no idea how most of this library _really_ works. I've just hacked at it to get it to work since @jschrewe abandoned django-mongodbforms, but it's a very complex piece of software with also a lot of bit-rot and stuff to get things for old python versions to work.

I've hoped that https://github.com/MongoEngine/django-mongoengine would pick up the slack, but so far it's still going slowly.

**Comment by lucashowell** [comment]
Hi @thomwiggers - thanks for looking into it and the quick response!

I saw your test case and tried to setup a simpler version of it in my environment (it's a pretty big codebase).  I'm having trouble trying to reproduce the issue consistently, and am now wondering if perhaps it's something else wrong in my codebase that upgrading mongodbforms / mongoengine has surfaced as a bit of a red herring.

Going to close for now - if I can consistently reproduce with a simple test case will re-open.  Thanks again for your help and taking on the onus of managing this library!  I agree it's surprisingly complex and hard to debug.

**Comment by lucashowell** [comment]
Hi @thomwiggers - this issue resurfaced for us today and I think i'm able to provide you an easy way to reproduce on your end.  The error occurs when  you have a DocumentForm for a Document that has a ReferenceField pointed at another Document with a self referential ReferenceField.  A bit of a mouthful, but I think that's a correct description.

This then causes a maximum recursion depth error in django-mongoengine-forms library when you try to import the DocumentForm in a django shell.

Simple example below.  To simulate the issue go into a django shell and try to import UserForm.

> python manage.py shell
> from collections import UserForm

```python

# collections.py

from mongoengine import Document
from mongoengine import StringField, ReferenceField

from mongodbforms import DocumentForm

class Group(Document):
    name = StringField(required=True, unique=True)
    parent_group = ReferenceField("Group")

class User(Document):
    name = StringField(required=True, unique=True)
    default_group = ReferenceField("Group")


class UserForm(DocumentForm):
    class Meta:
        document = User
        fields = ["name", "default_group", ]

```

**Comment by thomwiggers** [comment]
Ok so it turns out `LazyDocumentMetaWrapper` may call itself endlessly as its `_setup` will call `DocumentMetaWrapper` which will then try to access the `__class__` property of `LazyDocumentMetaWrapper` which then calls `_setup`. Great. As far as I can tell it shouldn't be doing lots of the operations that are happening on `LazyDocumentMetaWrapper`.

I've hacked a fix, I hope it helps. I don't really like this package all that much :disappointed:

**Comment by lucashowell** [comment]
@thomwiggers Awesome!  Thanks a bunch for the lightning fast turnaround on the fix.  We will try it out on our end.  Do you plan to tag it and repackage for PyPi anytime soon?

Thanks again for the help.

**Comment by thomwiggers** [comment]
I'm kinda scared that it'll break other things but so I'll have to test if it breaks my own application

**Comment by lucashowell** [comment]
Hi @thomwiggers just thought I'd touch base to see if you plan to tag and release a new version of django-mongoengine-forms that includes this fix anytime soon?

We've been running with this fix in a production environment that relies heavily on django-mongoengine-forms and seems to be running fine for us.

**Comment by thomwiggers** [comment]
I've released 0.4.4: https://pypi.org/project/django-mongoengine-forms/0.4.4/

**Comment by lucashowell** [comment]
@thomwiggers Awesome - thanks a bunch!
